rust/bin-shared/src/network_changes/windows.rs

```diff
@@ -81,61 +82,69 @@ use windows::{
 // async needed to match Linux
 #[allow(clippy::unused_async)]
 pub async fn new_dns_notifier(
-    _tokio_handle: tokio::runtime::Handle,
+    tokio_handle: tokio::runtime::Handle,
     _method: DnsControlMethod,
-) -> Result<async_dns::DnsNotifier> {
-    async_dns::DnsNotifier::new()
+) -> Result<Worker> {
+    Worker::new("Firezone DNS monitoring worker", move |tx, stopper_rx| {
+        async_dns::worker_thread(tokio_handle, tx, stopper_rx)?;
+        Ok(())
+    })
 }
 
+// Async on Linux due to `zbus`
+#[allow(clippy::unused_async)]
 pub async fn new_network_notifier(
     _tokio_handle: tokio::runtime::Handle,
     _method: DnsControlMethod,
-) -> Result<NetworkNotifier> {
-    NetworkNotifier::new().await
+) -> Result<Worker> {
+    Worker::new(
+        "Firezone network monitoring worker",
+        move |tx, stopper_rx| {
+            {
+                let com = ComGuard::new()?;
+                let _network_change_listener = Listener::new(&com, tx)?;
+                stopper_rx.blocking_recv().ok();
+            }
+            Ok(())
+        },
+    )
 }
 
-/// Notifies when we change Wi-Fi networks, change between Wi-Fi and Ethernet, or gain / lose Internet
-pub struct NetworkNotifier {
+/// Container for a worker thread that we can cooperatively stop.
+///
+/// The worker thread emits notifications with no data in them.
+pub struct Worker {
     inner: Option<WorkerInner>,
     rx: mpsc::Receiver<()>,
 }
 
-/// Needed so that `Drop` can consume the oneshot Sender and the thread's JoinHandle
-struct WorkerInner {
-    thread: std::thread::JoinHandle<Result<()>>,
-    stopper: tokio::sync::oneshot::Sender<()>,
+impl Drop for Worker {
+    fn drop(&mut self) {
+        self.close()
+            .expect("should be able to close WorkerInner cleanly");
+    }
 }
 
-impl NetworkNotifier {
-    // Async on Linux due to `zbus`
-    #[allow(clippy::unused_async)]
-    pub async fn new() -> Result<Self> {
+impl Worker {
+    fn new<
+        F: FnOnce(mpsc::Sender<()>, oneshot::Receiver<()>) -> Result<()> + Send + 'static,
```

Comment:
Style-nit: When bounds become more complex, I like to move them to `where` clauses. I find that rustfmt formats things nicer then. RA has a quick-action to do that.